python/mxnet/gluon/probability/block/stochastic_block.py

```diff
@@ -0,0 +1,130 @@
+        self._losscache = []
+        self._count = 0
+
+    def add_loss(self, loss):
+        self._count += 1
+        self._losscache.append(loss)
+
+    @staticmethod
+    def collectLoss(func):
+        """To accumulate loss during the forward phase, one could first decorate
+        hybrid_forward with `StochasticBlock.collectLos`s`,
+        and then collect the loss tensor `x` by calling self.add_loss(x).
+        For example, in the following forward function,
+        we generate samples from a Gaussian parameterized by `loc` and `scale` and
+        accumulate the KL-divergence between it and its prior into the block's loss storage.:
+        @StochasticBlock.collectLoss
+        def hybrid_forward(self, F, loc, scale):
+            qz = mgp.Normal(loc, scale)
+            # prior
+            pz = mgp.Normal(F.np.zeros_like(loc), F.np.ones_like(scale))
+            self.add_loss(mgp.kl_divergence(qz, pz))
+            return qz.sample()
+        """
+        @wraps(func)
+        def inner(self, *args, **kwargs):
+            # Loss from hybrid_forward
+            func_out = func(self, *args, **kwargs)
+            collected_loss = self._losscache
+            self._losscache = []
+            return (func_out, collected_loss)
+
+        return inner
+
+    def __call__(self, *args):
+        """Calls forward. Only accepts positional arguments."""
+        for hook in self._forward_pre_hooks.values():
+            hook(self, args)
+        self._losses = []
+        out = self.forward(*args)  # out[0]: net output, out[1]: collected loss
+        self._losses.extend(out[1])
+        for hook in self._forward_hooks.values():
+            hook(self, args, out)
+        return out[0]
+
+    @property
+    def losses(self):
+        return self._losses
+
+
+class StochasticSequential(StochasticBlock):
+    """Stack StochasticBlock sequentially.
+    """
+
+    def __init__(self, prefix=None, params=None):
+        super(StochasticSequential, self).__init__(
+            prefix=prefix, params=params)
+        self._layers = []
+
+    def add(self, *blocks):
+        """Adds block on top of the stack."""
+        for block in blocks:
+            self._layers.append(block)
+            self.register_child(block)
+
+    @StochasticBlock.collectLoss
+    def hybrid_forward(self, F, x):
+        for block in self._layers:
+            x = block(x)
+            if hasattr(block, '_losses'):
+                self.add_loss(block._losses)
+        return x
+
+    def __repr__(self):
+        s = '{name}(\n{modstr}\n)'
+        modstr = '\n'.join(['  ({key}): {block}'.format(key=key,
+                                                        block=_indent(block.__repr__(), 2))
+                            for key, block in self._children.items()])
+        return s.format(name=self.__class__.__name__,
+                        modstr=modstr)
+
+    def __getitem__(self, key):
+        layers = list(self._children.values())[key]
```

Comment:
Note the recent change to how HybridSequential stores the _children: https://github.com/apache/incubator-mxnet/pull/18376

You need to take that into account and dereference the weakref here